File: Task3/solution.cpp
#include <iostream>
#include <fstream>
#include <stack>
using namespace std;

int main()
{
    fstream file;
    string s;
    file.open("input.txt");
    file >> s;
    file.close();
    int n = s.length();
    stack<char> st;
    bool answer = true;
    for (int i = 0; i < n; i++)
    {
        if (s[i] == '(' || s[i] == '[' || s[i] == '{')
        {
            st.push(s[i]);
        }
        else if (!st.empty())
        {
            if (s[i] == ')' && st.top() == '(')
            {
                st.pop();
            }
            else
            {
                if (s[i] == ']' && st.top() == '[')
                {
                    st.pop();
                }
                else
                {
                    if (s[i] == '}' && st.top() == '{')
                    {
                        st.pop();
                    }
                    else
                    {
                        answer = false;
                        break;
                    }
                }
            }
        }
        else
        {
            answer = false;
            break;
        }
    }
    if (answer && st.size() == 0)
    {
        cout << "TRUE";
    }
    else
    {
        cout << "FALSE";
    }
    return 0;
}

File: Task3/solution.py
brackets = {')' : '(', ']' : '[', '}': '{'}

with open('input.txt') as file:
    s = file.readline()
    stack = list()
    for i in s:
        if i in brackets.values():
            stack.append(i)
        elif stack:
            if brackets[i] == stack[-1]:
                stack.pop()
            else:
                print('FALSE')
                break
        else:
            print('FALSE')
            break
    else:
        if len(stack) == 0:
            print('TRUE')
        else:
            print('FALSE')
